crates/sui-framework/packages/dwallet-system/sources/ethereum_state.move

```diff
@@ -0,0 +1,112 @@
+*/
+module dwallet_system::ethereum_state {
+    use dwallet::object::{Self, UID, ID};
+    use dwallet::transfer;
+    use dwallet::tx_context::TxContext;
+
+    /// Ethereum state object.
+    struct EthereumState has key {
+        id: UID,
+        /// serialised ConsensusStateManager
+        data: vector<u8>,
+        time_slot: u64,
+    }
+
+    /// Latest Ethereum state object.
+    /// Holds the ID of the latest Ethereum state object that is verified by the dWallet network.
+    struct LatestEthereumState has key, store {
+        id: UID,
+        eth_state_id: ID,
+        last_slot: u64,
+        eth_smart_contract_addr: vector<u8>,
+        eth_smart_contract_slot: u64,
+    }
+
+    /// Initializes the first Ethereum state with the given checkpoint.
+    /// Creates an EthereumState object, shares a LatestEthereumState object pointing to it,
+    /// and freezes the EthereumState object.
+    public fun init_state(
+        checkpoint: vector<u8>,
+        network: vector<u8>,
+        eth_smart_contract_addr: vector<u8>,
+        eth_smart_contract_slot: u64,
+        ctx: &mut TxContext
+    ) {
+        let state_data = create_initial_eth_state_data(checkpoint, network);
+        let state = EthereumState {
+            id: object::new(ctx),
+            data: state_data,
+            time_slot: 0u64,
+        };
+
+        transfer::share_object(LatestEthereumState {
+            id: object::new(ctx),
+            eth_state_id: object::id(&state),
+            last_slot: state.time_slot,
+            eth_smart_contract_addr,
+            eth_smart_contract_slot,
+        });
+        transfer::freeze_object(state);
+    }
+
+    /// Updates the latest Ethereum state reference if the provided EthereumState object has a newer time slot.
+    public fun update_latest_eth_state(
+        self: &mut LatestEthereumState,
+        eth_state: &EthereumState,
+    ) {
+        if (eth_state.time_slot > self.last_slot) {
+            self.eth_state_id = object::id(eth_state);
+            self.last_slot = eth_state.time_slot;
+        }
+    }
+
+    /// Verifies the new Ethereum state according to the provided updates, and updates the LatestEthereumState object
+    /// if the new state is valid and has a newer time slot.
+    public fun verify_new_state(
+        updates_bytes: vector<u8>,
+        finality_update_bytes: vector<u8>,
+        optimistic_update_bytes: vector<u8>,
+        state_bytes: vector<u8>,
+        latest_ethereum_state: &mut LatestEthereumState,
+        ctx: &mut TxContext,
+    ) {
+        let (data, time_slot) = verify_eth_state(
+            updates_bytes,
+            finality_update_bytes,
+            optimistic_update_bytes,
+            state_bytes
+        );
+
+        let new_state = EthereumState {
+            id: object::new(ctx),
+            data,
+            time_slot,
+        };
+
+        if (new_state.time_slot > latest_ethereum_state.last_slot) {
```

Comment:
Also here check networks match

Reply:
I do. Let's try to find how this is enforced in `helios`.

Reply:
When you say it is "pulled" using the ID from `LatestEthereumState`, this implies you are trusting the user honestly followed protocol. But there is no reason for they to do so, nor for you to trust so. Trust nothing, verify everything.